onnxruntime/core/framework/tensor.cc

```diff
@@ -8,25 +8,26 @@
 using namespace std;
 namespace onnxruntime {
 
-Tensor::Tensor(MLDataType p_type,
-               const TensorShape& shape,
-               BufferNakedPtr p_data,
-               const OrtAllocatorInfo& alloc,
-               AllocatorPtr deleter,
-               const int64_t offset)
+Tensor::Tensor(MLDataType p_type, const TensorShape& shape, BufferNakedPtr p_data, const OrtAllocatorInfo& alloc,
+               int64_t offset)
     : alloc_info_(alloc) {
   ORT_ENFORCE(p_type != nullptr);
-  Init(p_type, shape, p_data, alloc, std::move(deleter), offset);
+  Init(p_type, shape, p_data, nullptr, offset);
 }
 
-void Tensor::Init(MLDataType p_type,
-                  const TensorShape& shape,
-                  void* p_raw_data,
-                  const OrtAllocatorInfo& alloc,
-                  AllocatorPtr deleter,
-                  const int64_t offset) {
-  if (shape.Size() < 0)
-    throw std::runtime_error("shape.Size() must >=0");
+Tensor::Tensor(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator, int64_t offset)
+    : alloc_info_(allocator->Info()) {
+  ORT_ENFORCE(p_type != nullptr);
+  int64_t shape_size = shape.Size();
+  if (shape_size < 0 || static_cast<uint64_t>(shape_size) >= std::numeric_limits<size_t>::max())
+    ORT_THROW("shape.Size() must >=0");
+  void* p_data = allocator->AllocArray(static_cast<size_t>(shape_size), p_type->Size());
+  Init(p_type, shape, p_data, allocator, offset);
+}
```

Comment:
I'd be inclined to do the allocation Init if p_raw_data is nullptr so there's only one check of size >= 0.